src/main/java/com/puppycrawl/tools/checkstyle/checks/TrailingCommentCheck.java

```diff
@@ -200,76 +188,100 @@ public boolean isCommentNodesRequired() {
+     *
+     * @param ast Detail ast element to be checked.
+     */
+    private void checkSingleLineComment(DetailAST ast) {
         final int lineNo = ast.getLineNo();
-        if (illegalLines.get(lineNo) != null
-                && illegalLines.get(lineNo).equals(ast.getColumnNo())) {
+        final String comment = ast.getFirstChild().getText();
+        final String line = getLines()[lineNo - 1];
+        final String lineBefore = line.substring(0, ast.getColumnNo());
+
+        if (!format.matcher(lineBefore).find()
+                && !checkSingleLine(comment)) {
             log(ast, MSG_KEY);
         }
     }
 
-    @Override
-    public void beginTree(DetailAST rootAST) {
-        final Map<Integer, TextBlock> cppComments = getFileContents()
-                .getSingleLineComments();
-        final Map<Integer, List<TextBlock>> cComments = getFileContents()
-                .getBlockComments();
-        final Set<Integer> lines = new HashSet<>();
-        lines.addAll(cppComments.keySet());
-        lines.addAll(cComments.keySet());
+    /**
+     * Method for block comment.
+     *
+     * @param ast Detail ast element to be checked.
+     */
+    private void checkBlockComment(DetailAST ast) {
+        final int lineNo = ast.getLineNo();
+        final String comment = ast.getFirstChild().getText();
+        String line = getLines()[lineNo - 1];
 
-        for (Integer lineNo : lines) {
-            final String line = getLines()[lineNo - 1];
-            final String lineBefore;
-            final TextBlock comment;
-            if (cppComments.containsKey(lineNo)) {
-                comment = cppComments.get(lineNo);
-                lineBefore = line.substring(0, comment.getStartColNo());
-            }
-            else {
-                final List<TextBlock> commentList = cComments.get(lineNo);
-                comment = commentList.get(commentList.size() - 1);
-                lineBefore = line.substring(0, comment.getStartColNo());
+        if (line.length() > ast.getLastChild().getColumnNo() + 1) {
+            line = line.substring(ast.getLastChild().getColumnNo() + 2);
+        }
+        final Pattern removeSlash = Pattern.compile("/");
+        line = removeSlash.matcher(line).replaceAll("");
 
-                // do not check comment which doesn't end line
-                if (comment.getText().length == 1
-                        && !CommonUtil.isBlank(line
-                            .substring(comment.getEndColNo() + 1))) {
-                    continue;
-                }
-            }
-            if (!format.matcher(lineBefore).find()
-                && !isLegalComment(comment)) {
-                illegalLines.put(lineNo, comment.getStartColNo());
-            }
+        final String lineBefore = getLines()[lineNo - 1].substring(0, ast.getColumnNo());
+
+        // do not check comment which doesn't end line
+        if (!(ast.getLineNo() == ast.getLastChild().getLineNo()
```

Comment:
it is better to avoid negation of whole expression. It seems to me that it can be changed - all conditions inside are inverted

Reply:
Done